### ISSUE TITLE
self._computed role and undesired behaviour

### ISSUE THREAD

**Issue by astrocyted** [Lightning-AI/torchmetrics#1841]
Subclassing the Metric class in my custom metric class im collecting model predictions and at the end of the epoch I calculate  multiple metrics using those predictions.  for this I am passing an argument to the compute() method to just use the stored predictions and calculate metrics. however after calling compute once, the successive compute() cvalls are blocked due to the flag self._computed being set to previous computed value after first call apparently which prevents from any further computations. Is there a good explanation for this whole _computed flag ?

Of course I "can" create different metric classes for different metrics that all use the same prediction scores, but that would be quite stupid and redundant as all metrics are using the exact same data and that would just accumulate multiple copies of a same data. So to me ._computed just seems unnecssarily blocking.

**Comment by SkafteNicki** [comment]
Hi @astrocyted, thanks for raising this issue.

There are a couple of reasons:
* To prevent that repeated calls to `compute` leads to repeated computations
* We never had a use case where people have arguments in `compute`, thus calling `compute` would always return the same value and therefore it makes sense to cache
* For compatibility reasons with the logging in pytorch lightning

All that said, we had another usecase recently where this caching behaviour was also a problem, and that lead us to introduce a new global argument to all base metrics. Its called `compute_with_cache` and introduced in PR https://github.com/Lightning-AI/torchmetrics/pull/1754. By default it is `True` and computed results are caches. If setting to `False` then they are not cache.

Please see the following small example:

```python
from torchmetrics.metric import Metric
import torch

class M(Metric):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.add_state("x", default=torch.tensor(0), dist_reduce_fx="sum")

    def update(self, x):
        self.x += x

    def compute(self, pow=2):
        return self.x ** pow

m = M()
m.update(torch.tensor(2))
print(m.compute(pow=2))  # prints 4 (2^2)
print(m.compute(pow=3))  # prints 4 due to cache

m = M(compute_with_cache=False)
m.update(torch.tensor(2))
print(m.compute(pow=2))  # prints 4 (2^2)
print(m.compute(pow=3))  # prints 8 (2^3)
```

To get access to the new feature please update to master:
```
pip install https://github.com/Lightning-AI/torchmetrics/archive/master.zip
```

**Comment by astrocyted** [comment]
Thanks @SkafteNicki for the explanation and the solution. For now, I'm alternatively providing a list of arguments to compute() and compute all the desired metrics in one call, circumventing the cache issue, although now i have an issue with my internal state variables not being reset automatically after calling compute().

However I genrally find there is something missing in the base class for cases when multiple metrics are computed (say in obj. detection) using essentially same model predictions and just instantiating a IoU, gIoU, ... only for all of them to independently accumulate predictions. for now Im like i said writing a wrapper to compute all metrics at once using single set of state variable but it would be nice if that was an option generally to integrate all desired metrics into one object smoothly without having to instantiate separate Metric class for each.

**Comment by SkafteNicki** [comment]
> Thanks @SkafteNicki for the explanation and the solution. For now, I'm alternatively providing a list of arguments to compute() and compute all the desired metrics in one call, circumventing the cache issue, although now i have an issue with my internal state variables not being reset automatically after calling compute().

@astrocyted Not sure what is causing the problem with the internal state variables not being reset (would need to see the code for that), but if you need custom logic regarding the reset, then you can manually overwrite reset

```python
def reset(self):
     super().reset()
     self._manual_reset_stuff()
```
 
> However I genrally find there is something missing in the base class for cases when multiple metrics are computed (say in obj. detection) using essentially same model predictions and just instantiating a IoU, gIoU, ... only for all of them to independently accumulate predictions. for now Im like i said writing a wrapper to compute all metrics at once using single set of state variable but it would be nice if that was an option generally to integrate all desired metrics into one object smoothly without having to instantiate separate Metric class for each.

We have that functionality in the [MetricCollection object](https://torchmetrics.readthedocs.io/en/stable/pages/overview.html#metriccollection). If I have metric M1 and metric M2, and both of them share the exact same internal states, `MetricCollection` will automatically make sure that if I call `.update` on the collection, then only `update` of M1 is called and the resulting state is propergated to M2. As an example, here is `Precision` and `Recall` which have the same state but differs in their compute
```python
from torchmetrics import MetricCollection
from torchmetrics.classification import BinaryPrecision, BinaryRecall
import torch

preds = lambda : torch.rand(10)
target = lambda : torch.randint(2, (10,))

m = MetricCollection([BinaryPrecision(), BinaryRecall()])
print(m.compute_groups)  # {0: ['BinaryPrecision'], 1: ['BinaryRecall']}
m.update(preds(), target())
print(m.compute_groups)  # {0: ['BinaryPrecision', 'BinaryRecall']}
m.update(preds(), target())
res = m.compute()
```
as you can see in the example, the `compute_groups` property indicates if two metrics share state. Initially, we do not know this and precision and recall is in individual groups. After the first `update` call we check if states matches (by name and value), and then merge precision and recall into the same group because they match. Then calls to `update` afterwards will only call the `update` method of the first metric in each group and propagate the state to the rest of the group.
